rts/System/Sync/SyncChecker.cpp

```diff
@@ -12,6 +12,41 @@
 unsigned CSyncChecker::g_checksum;
 int CSyncChecker::inSyncedCode;
 
+#ifdef SYNC_HISTORY
+
+unsigned CSyncChecker::nextHistoryIndex = 0;
+unsigned CSyncChecker::nextFrameIndex = 0;
+unsigned CSyncChecker::logs[MAX_SYNC_HISTORY];
+unsigned CSyncChecker::logFrames[MAX_SYNC_HISTORY_FRAMES];
+
+void CSyncChecker::NewGameFrame()
+{
+	logFrames[nextFrameIndex++] = nextHistoryIndex;
+	if (nextFrameIndex == MAX_SYNC_HISTORY_FRAMES)
+		nextFrameIndex = 0;
+}
+
+void CSyncChecker::LogHistory()
+{
+	logs[nextHistoryIndex++] = g_checksum;
+	if (nextHistoryIndex == MAX_SYNC_HISTORY)
+		nextHistoryIndex = 0;
+}
+
+std::tuple<unsigned, unsigned, unsigned*> CSyncChecker::GetFrameHistory(unsigned rewindFrames)
+{
+	int endFrameIndex = nextFrameIndex - rewindFrames;
+	int startFrameIndex = endFrameIndex - 1;
+
+	if (endFrameIndex < 0)
+		endFrameIndex = MAX_SYNC_HISTORY_FRAMES + endFrameIndex;
+	if (startFrameIndex < 0)
+		startFrameIndex = MAX_SYNC_HISTORY_FRAMES + startFrameIndex;
```

Comment:
Maybe put together a specialized templated RingBuffer class for this and similar cases?